### PR TITLE
Updated SQLite Historizing

### PR DESCRIPTION
**Pull request by zerox1212**
I tried to fix everything you told me about in https://github.com/FreeOpcUa/python-opcua/pull/155

I don't know how you want to organize the code. I also don't know how the user should choose which history method, for now I just commented out the SQLite import.

Perhaps you can just take history_sql.py into the main branch and then reorganize the code how you see fit.

### CODE REVIEW

File: opcua/server/history.py [hunk 1]
```diff
@@ -90,16 +37,17 @@ def save_node_value(self, node, datavalue):
             data = data[-count:]
 
     def read_node_history(self, node, start, end, nb_values):
+        node_id = node.NodeId
```
Comment (destogl) [inline opcua/server/history.py]: Is this spelled wrongly? Since type `Node` has `nodeid` attribute?

File: opcua/server/history.py [hunk 2]
```diff
@@ -4,83 +4,30 @@
...
         self._datachanges_period = {}
         self._events = {}
 
     def new_historized_node(self, node, period, count=0):
-        self._datachanges[node] = []
-        self._datachanges_period[node] = period, count
-
-    def new_historized_event(self, event, period):
-        self._events = []
+        node_id = node.nodeid
```
Comment (destogl) [inline opcua/server/history.py]: I would remove this line since IMHO is not important it also doesn't really remove complexity

File: opcua/server/history.py [hunk 3]
```diff
@@ -4,83 +4,30 @@
...
+        self._datachanges[node_id] = []
+        self._datachanges_period[node_id] = period, count
 
     def save_node_value(self, node, datavalue):
-        data = self._datachanges[node]
-        period, count = self._datachanges_period[node]
+        node_id = node.nodeid
```
Comment (destogl) [inline opcua/server/history.py]: The same as line 23

**Comment by destogl** [comment]
For me it looks good. I would just put  `history_interface.py`  file in  `history.py` since this would be more Pythonic way. `history_sql.py` should stay since it needs additional module etc...

Anyway, nice work!

**Comment by oroulet** [comment]
> I would just put history_interface.py file in history.py
agree. Can you do this @zerox1212 ?

File: opcua/server/history.py (hunk 1 above)
Comment (oroulet) [inline opcua/server/history.py]: yes it is very strange.... I will  pull that request and check these things

File: opcua/server/history.py (hunk 2 above)
Comment (oroulet) [inline opcua/server/history.py]: agree. Can you do this @zerox1212 ?

File: opcua/server/history.py (hunk 1 above)
Comment (zerox1212) [inline opcua/server/history.py]: It's because in that part you don't get a UA Node. You get a ua.HistoryReadValueId object. In that object node id is in camel case.

Node object spelling:
image

HistoryReadValueId spelling:
image

**Comment by zerox1212** [comment]
I can move the interface back to history.py, however how can I import my `HistorySQlite` class from history_sql.py then?

The only way it works is if `from opcua.server.history_sql import HistorySQLite` is after the class definition of the interface. I thought all import statements should be at the top?

File: opcua/server/history_sql.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            return 'Q'
+        elif variant_code is 10:  # Float
+            return 'f'
+        elif variant_code is 11:  # Double
+            return 'd'
+        elif variant_code in (12,):  # String
+            return "s"
+        else:
+            # FIXME: Should raise exception here that historizing of node type isn't supported in SQL storage interface
+            return None
```
Comment (zerox1212) [inline opcua/server/history_sql.py]: Do you guys have a recommendation on how to handle the case where the node being historized can't be packed into a binary for SQL?

File: opcua/server/history.py (hunk 1 above)
Comment (destogl) [inline opcua/server/history.py]: Can you maybe than put different name of the parameter to be clearer

File: opcua/server/history_sql.py (hunk 4 above)
Comment (destogl) [inline opcua/server/history_sql.py]: You should here check the constants not the hard-coded number, I would write e.g.:

```
...
variant_type is VariantType.Boolean:
     return '?'
...
```

By the way, there is solution to do this with [dictionary](https://bytebaker.com/2008/11/03/switch-case-statement-in-python/) which could be possibly cleaner since you have constants... @oroulet: which way do you prefere?

I have no proposal for your problem sorry....

File: opcua/server/history.py (hunk 1 above)
Comment (oroulet) [inline opcua/server/history.py]: Why do you pass the entire object if you only use the nodeid?

**Comment by destogl** [comment]
> I can move the interface back to history.py, however how can I import my HistorySQlite class from history_sql.py then?

> The only way it works is if from opcua.server.history_sql import HistorySQLite is after the class definition of the interface. I thought all import statements should be at the top? HistorSQLite can't import the interface class before it's declared.

this is all correct you wrote, but I don't understand your problem, since you just move `HistoryStorageInterface` class and all imports stay the same.... or am I missing something?

File: opcua/server/history_sql.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        try:
+            _c_new.execute('CREATE TABLE "{tn}" (ServerTimestamp TIMESTAMP,'
+                           ' SourceTimestamp TIMESTAMP,'
+                           ' StatusCode INTEGER,'
+                           ' Value TEXT,'
+                           ' VariantType INTEGER,'
+                           ' ValueBinary BLOB)'.format(tn=node_id))
+
+        except sqlite3.Error as e:
+            print(node_id, 'Historizing SQL Table Creation Error:', e)
```
Comment (oroulet) [inline opcua/server/history_sql.py]: add a logger object: self.logger = logging.getLogger(__name__) (in __init__)
then
self.logger.error("Creating table failed %", ex)

**Comment by zerox1212** [comment]
See my example:
history.py
```python
from opcua.server.history_sql import HistorySQLite

class HistoryStorageInterface(object):
    code....
```

history_sql.py
```python
from opcua.server.history import HistoryStorageInterface

class HistoryStorageSQL(HistoryStorageInterface)
    code...
```

This won't work because HistorySQLite will be imported before the class HistoryStorageInterface is declared.

File: opcua/server/history_sql.py (hunk 5 above)
Comment (zerox1212) [inline opcua/server/history_sql.py]: Yes I wanted to ask you about this. Should there be a logger per backend? Or should there be a "history" logger that all backends use?

File: opcua/server/history_sql.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            return "s"
+        else:
+            # FIXME: Should raise exception here that historizing of node type isn't supported in SQL storage interface
+            return None
+
+    def _get_table_name(self, node):
+        if type(node) is Node:
+            return str(node.nodeid.NamespaceIndex) + '_' + str(node.nodeid.Identifier)
+        if type(node) is ua.HistoryReadValueId:
+            return str(node.NodeId.NamespaceIndex) + '_' + str(node.NodeId.Identifier)
```
Comment (oroulet) [inline opcua/server/history_sql.py]: we may have an issue here. Identifier can be very strange things such as guid... but we can skip that issue for now...

File: opcua/server/history_sql.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+        value_blob = self._pack_value(datavalue)
+
+        # insert the data change into the database
+        try:
+            _c_sub.execute('INSERT INTO "{tn}" VALUES (?, ?, ?, ?, ?, ?)'.format(tn=node_id), (datavalue.ServerTimestamp,
+                                                                                               datavalue.SourceTimestamp,
+                                                                                               datavalue.StatusCode.value,
+                                                                                               str(datavalue.Value.Value),
+                                                                                               datavalue.Value.VariantType.value,
```
Comment (oroulet) [inline opcua/server/history_sql.py]: what about saving VariantType.name instead of value? I think that it does not impact much performance and make things more readable. We can always change it later. just a suggestion...

File: opcua/server/history_sql.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+    def save_event(self, event):
+        raise NotImplementedError
+
+    def read_event_history(self, start, end, evfilter):
+        raise NotImplementedError
+
+    def _pack_value(self, datavalue):
+        variant_code = datavalue.Value.VariantType.value
+        return struct.pack(self._get_pack_type(variant_code), datavalue.Value.Value)
```
Comment (oroulet) [inline opcua/server/history_sql.py]: OK, this is not necessary. If you want to save variant as binary ust called Variant.to_binary() and variant.from_binary(). No need to use pack directly. 

Would be nice to save data as a python type, but we will get issue with extensions objects, etc,,,,

File: opcua/server/history_sql.py (hunk 4 above)
Comment (oroulet) [inline opcua/server/history_sql.py]: As written in last comment this method is not necessary. remove it and use to_binary()

File: opcua/server/history_interface.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        raise NotImplementedError
+
+    def save_node_value(self, node, datavalue):
+        """
+        Called when the value of a historized node has changed and should be saved in history
+        Returns None
+        """
+        raise NotImplementedError
+
+    def read_node_history(self, node, start, end, nb_values):
```
Comment (oroulet) [inline opcua/server/history_interface.py]: why did you change passed type from nodeid to node? it does not look like you are using it anywhere

**Comment by oroulet** [comment]
Sorry for all the comments, I know it is hard to understand the entire UA sprotocol and this particular stack

File: opcua/server/history_sql.py (hunk 7 above)
Comment (zerox1212) [inline opcua/server/history_sql.py]: I will do it if I don't need to convert it to a value manually...

Can I pass a string to VariantType()? I am not sure how Enums work. I assumed I needed to pass a value to it.

File: opcua/server/history_sql.py (hunk 8 above)
Comment (zerox1212) [inline opcua/server/history_sql.py]: I tried this several times. I couldn't get it to work because `Variant.from_binary()` expects a buffer. SQL returns a byte array.

File: opcua/server/history_interface.py (hunk 9 above)
Comment (zerox1212) [inline opcua/server/history_interface.py]: Sorry I forgot to change that back. Before I was using the node name for table naming. Now I only use the nodeid so I will change it back.

**Comment by zerox1212** [comment]
I appreciate all the comments. I try my best to help you guys, but it seems like the amount of time you spend looking at my could you could of written it yourself. Sorry for that.

**Comment by oroulet** [comment]
It is faster to comment than to write :-) and then you learn and maybe
write more  code later !

File: opcua/server/history_sql.py (hunk 5 above)
Comment (oroulet) [inline opcua/server/history_sql.py]: Make your own for your SQL backbend

File: opcua/server/history_sql.py (hunk 8 above)
Comment (oroulet) [inline opcua/server/history_sql.py]: Wrap your byte array with opcua.utils.Buffer then it will work

File: opcua/server/history.py [hunk 10]
```diff
@@ -4,7 +4,70 @@
...
+    def stop(self):
+        """
+        Called when the server shuts down
+        Can be used to close database connections etc.
+        """
+        raise NotImplementedError
+
 
 # if you want to use an SQL based history uncomment this import and change the storage type of the history manager
 # from opcua.server.history_sql import HistorySQLite
```
Comment (zerox1212) [inline opcua/server/history.py]: @destogl This is why I moved the interface to it's own file. I don't know the correct way to import the HistorySQLite backend, but you can't do it at the top of history.py.

**Comment by zerox1212** [comment]
I did all the corrections you spoke of except using `Variant.to_binary()` and `Variant.from_binary()`. This because I don't know how to get `from_binary() to work as it requires a buffer object. SQL doesn't return that when I SELECT it.

Please check it, thanks.

**Comment by oroulet** [comment]
To use from_binary from a string: xx.from_binary(opcua.utils.Buffer(mystring))

**Comment by zerox1212** [comment]
I will give that a try and test it if it works.

I also saw some people using `sqlite3.Binary(blob)` when inserting into the database so I need to look into what that does.

**Comment by zerox1212** [comment]
I implemented `to_binary()` and `from_binary()` as you mentioned and it is working. Thanks for that.

**Comment by destogl** [comment]
> I appreciate all the comments. I try my best to help you guys because I really like this project, but it seems like the amount of time you spend looking at my code you could of written it yourself. Sorry for that.

Don't worry about that, all of us need feedback about the code, this is what make the code nice. You are welcome to comment my PRs as well :)

> See my example: ...

Thanks for example now I understand the problem, sorry I missed that. But I am not sure that this is the best way for doing that, so to say to import `HistoryStorageSQL` from `history.py`. I would propose the following:

* The `history_sql.py ` should be imported from server implementation and not from `history.py` file. So we have always dict history as default and if user wants to use sql than set it by himself so in the example file `server-minimal-history-sql` i would write following:

```
from opcua.server.history_sql import HistorySQLite
...
server = Server()
server.history_manager.set_storage(HistorySQLite())
...
```

@oroulet: should we have option to enable/disable history? I can imagine use cases where user doesn't want to use history...

**Comment by zerox1212** [comment]
Having a set_storage() method works for me. If we could clean up these last points and commit then I can move forward and take a look at historizing events.

Thanks everyone.

**Comment by oroulet** [comment]
@zerox1212 look at #158 . I moved your work there. and I close this one
